feat: Add metadata to report

briandonahue opened the pull request:
**Summary:**

Closes #1363.

Add a metadata / overview section to the HTML report.

**Expected behavior:** 

Explain and/or show screenshots for how you expect the pull request to work in your testing (in case other devices exhibit different behavior).


Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

briandonahue commented:
@isabelle-dr This is still a work in progress, and obviously the report needs better organization, but wanted to show you the data as it's currently being output and see if you had any specific requests for organizing it on the report:

<img width="905" alt="image" src="https://user-images.githubusercontent.com/40084/232857097-315f699a-ea8a-4543-bff2-93b80800db32.png">

For the "presence of spec features" section, how do you envision that looking?

briandonahue commented:
Updated Screen shot: 
<a href="https://user-images.githubusercontent.com/40084/232911218-d3748ff6-3575-4e11-b5c4-77a94e5f53bc.png" target="_blank"><img width="600" alt="image" src="https://user-images.githubusercontent.com/40084/232911218-d3748ff6-3575-4e11-b5c4-77a94e5f53bc.png"></a>

briandonahue commented:
@bdferris-v2 @davidgamez Leaving this a draft for now, but looking for any feedback/suggestions on the approach and/or the report organization/layout

isabelle-dr commented:
As linked to the GitHub issue, [here is a scoping doc](https://docs.google.com/document/d/157X9lMTkZHH7oYxK1fYPiY-rMh_VWyPMW_s9wImOP-I/edit#) for this PR.

briandonahue commented:
Getting a little closer.  May need to call in some CSS support for a more responsive design :) 
<img width="1522" alt="image" src="https://user-images.githubusercontent.com/40084/234622744-50cac9c3-3be7-4a12-8a96-eeda724333d1.png">

isabelle-dr commented:
This is beautiful (and why so many shapes in this dataset? 😅)

briandonahue commented:
@isabelle-dr 

> why so many shapes in this dataset? 😅 is 1171 the number of unique shape_ids or the number of records?

Yes!  I fixed it.  Are all the entities under `Counts` supposed to count unique IDs?  Right now only Blocks and Shapes are doing so.

> You can remove the section at the end (starting at "parameters used")

Yes!


I also have updated the HTML with the help of my BFF ChatGPT. The bottom (original) compliance report is no longer affected by the Summary CSS, and the Summary columns stack as you narrow the browser window.  

Full width:

<img width="1356" alt="image" src="https://user-images.githubusercontent.com/40084/236015076-abda4bfa-d556-4315-887d-db4ec26a3526.png">

Narrower width:

<img width="803" alt="image" src="https://user-images.githubusercontent.com/40084/236015274-a52bc805-9f75-4bca-8100-18cfdfe072fc.png">

It is even pretty viewable on a mobile-size width, but the bottom table currently does not collapse so it will require horizontal scrolling 😄 

I think this is now ready for review!

isabelle-dr commented:
How nice @briandonahue (and thank you Chat GPT)!

> Are all the entities under Counts supposed to count unique IDs? Right now only Blocks and Shapes are doing so.

Yes, we want the # of unique ID fields for the counts. It shouldn't make a difference for the others because each record should have a unique ID, but we could have duplicates so I'd rather look at the unique IDs all the time. 

Why are we having all these corrupted sources in the acceptance test report all of a sudden? cc @davidgamez

briandonahue commented:
> Yes, we want the # of unique ID fields for the counts. It shouldn't make a difference for the others because each record should have a unique ID, but we could have duplicates so I'd rather look at the unique IDs all the time.

@isabelle-dr I've updated it to always count unique IDs and simplified the code a bit to use generics (@davidgamez open to feedback on this if there is a cleaner, more concise approach).

isabelle-dr commented:
@briandonahue we're getting there!
In order to test this PR, could you share the dataset you've been using to generate these reports?

Another question I have is: under the "Files included", are you listing all of them or only the official GTFS files? 
Maybe it wasn't clear in the Requirements doc, we want all files listed there.

davidgamez commented on main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:
```diff
@@ -0,0 +1,131 @@
+
+  public ArrayList<AgencyMetadata> agencies = new ArrayList<>();
+
+  public static FeedMetadata from(GtfsFeedContainer feedContainer) {
```
suggestion: this method populates all the necessary fields and counts. I suggest making the class-default constructor protected to avoid instantiation and prevent unexpected behaviors while calling methods in the class without proper field initialization.

davidgamez commented on main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java:
```diff
@@ -121,14 +118,17 @@ public Status run(ValidationRunnerConfig config) {
       feedContainer =
           loadAndValidate(
               validatorLoader, feedLoader, noticeContainer, gtfsInput, validationContext);
+      // TODO: collect metadata from feedContainer and include in report.
+
```
question: is this been addressed in line 127?

davidgamez reviewed (approved):
LGTM!

briandonahue commented:
> @briandonahue we're getting there! In order to test this PR, could you share the dataset you've been using to generate these reports?

Yes, these were 2 I was testing with:
[Westchester_County_Bee_Line.zip](https://github.com/MobilityData/gtfs-validator/files/11490480/Westchester_County_Bee_Line.zip) 
[alhambra-ca-us.zip](https://github.com/MobilityData/gtfs-validator/files/11490473/alhambra-ca-us.zip)

> Another question I have is: under the "Files included", are you listing all of them or only the official GTFS files? Maybe it wasn't clear in the Requirements doc, we want all files listed there.

I was not, but I've updated it now to do so.

/cc @davidgamez

isabelle-dr commented:
@briandonahue can you share a screenshot of your latest version for [alhambra-ca-us.zip](https://github.com/MobilityData/gtfs-validator/files/11490473/alhambra-ca-us.zip)?

briandonahue commented:
> @briandonahue can you share a screenshot of your latest version for [alhambra-ca-us.zip](https://github.com/MobilityData/gtfs-validator/files/11490473/alhambra-ca-us.zip)?

@isabelle-dr Here you go 😄 

<img width="1260" alt="image" src="https://github.com/MobilityData/gtfs-validator/assets/40084/bd537bf7-4c9d-49b0-aa11-40ef3e51c3ac">

isabelle-dr commented:
Perfect! Merging
